src/main/java/com/hubspot/jinjava/objects/date/StrftimeFormatter.java

```diff
@@ -16,101 +22,90 @@ public class StrftimeFormatter {
+        .build();
+
+    NOMINATIVE_COMPONENTS =
+      ImmutableMap
+        .<Character, StrftimeConversionComponent>builder()
+        .put('B', pattern("LLLL"))
+        .build();
   }
 
   /**
-   * Parses a string in python strftime format, returning the equivalent string in java date time format.
+   * Build a {@link DateTimeFormatter} that matches the given Python <code>strftime</code> pattern.
    *
-   * @param strftime
-   * @return date formatted as string
+   * @see <a href="https://strftime.org/">Python <code>strftime</code> cheatsheet</a>
    */
-  public static String toJavaDateTimeFormat(String strftime) {
+  public static DateTimeFormatter toDateTimeFormatter(String strftime) {
     if (!StringUtils.contains(strftime, '%')) {
-      return strftime;
+      return DateTimeFormatter.ofPattern(strftime);
     }
 
-    StringBuilder result = new StringBuilder();
+    DateTimeFormatterBuilder builder = new DateTimeFormatterBuilder();
 
     for (int i = 0; i < strftime.length(); i++) {
       char c = strftime.charAt(i);
-      if (c == '%' && strftime.length() > i + 1) {
-        c = strftime.charAt(++i);
-        boolean stripLeadingZero = false;
-        String[] conversions = CONVERSIONS;
+      if (c != '%' || strftime.length() <= i + 1) {
+        builder.appendLiteral(c);
+        continue;
+      }
 
-        if (c == '-') {
-          stripLeadingZero = true;
-          c = strftime.charAt(++i);
-        }
+      c = strftime.charAt(++i);
+      boolean stripLeadingZero = false;
+      Map<Character, StrftimeConversionComponent> components = COMPONENTS;
 
-        if (c == 'O') {
-          c = strftime.charAt(++i);
-          conversions = NOMINATIVE_CONVERSIONS;
-        }
+      if (c == '-') {
+        stripLeadingZero = true;
+        c = strftime.charAt(++i);
+      }
 
-        if (c > 255) {
-          // If the date format has invalid character that is > ascii (255) then
-          // maintain the behaviour similar to invalid ascii char <= 255 i.e. append null
-          result.append(conversions[0]);
-        } else {
-          if (stripLeadingZero) {
-            result.append(conversions[c].substring(1));
-          } else {
-            result.append(conversions[c]);
-          }
-        } // < 255
-      } else if (Character.isLetter(c)) {
-        result.append("'");
-        while (Character.isLetter(c)) {
-          result.append(c);
-          if (++i < strftime.length()) {
-            c = strftime.charAt(i);
-          } else {
-            c = 0;
-          }
-        }
-        result.append("'");
-        --i; // re-consume last char
-      } else {
-        result.append(c);
+      if (c == 'O') {
+        c = strftime.charAt(++i);
+        components = NOMINATIVE_COMPONENTS;
       }
+
+      Optional
+        .ofNullable(components.get(c))
+        .orElseThrow(() -> new InvalidDateFormatException(strftime))
+        .append(builder, stripLeadingZero);
```

Comment:
Just to make sure I'm understanding this right:

The previous behavior when there was an invalid date character was to no-op but now we are a throwing an error?

Reply:
> The previous behavior when there was an invalid date character was to no-op but now we are a throwing an error?

Good eye not sure why I didn't think about this before. You'd be right if this class was being used directly, but most usages of it will be through `Jinjava.render` which catches the error. [I added a test](https://github.com/HubSpot/jinjava/pull/930/files/cd7c4169afd8f5be29c2370fe95c2c189aa8071b..dec4d209792f1719ad6e8f8a1908d43c27d5d319) in the PR before this one to confirm the current behavior.

> ... could we specify which character is invalid in the exception message, so a developer can more easily debug the issue?

Also a good point. I added that context to the error message which you can see in [the updated version of the aforementioned test](https://github.com/HubSpot/jinjava/blob/c6065e7e18a7fae54613ff26bfa53eeab856124a/src/test/java/com/hubspot/jinjava/lib/filter/DateTimeFormatFilterTest.java#L169).

> We could also potentially link a dev to docs for all the valid characters they could use in a format string

Since this is in Jinjava and not the wrapping HubL implementation, I don't think it makes sense to link to HubSpot developer docs here.

Reply:
> Since this is in Jinjava and not the wrapping HubL implementation, I don't think it makes sense to link to HubSpot developer docs here.

ah :good-point: